File: codility/py/13-ladder.py
# You have to climb up a ladder. The ladder has exactly N rungs, numbered from 1 to N. With each step, you can ascend by one or two rungs. More precisely:

# combos calculates combinations of 2 for provided v and stores in memo
def combos(v, memo):
    if v < 0:
        return 0  # Invalid
    while len(memo) <= v:
        # Calculate all values up to memo[v]
        # If v < len(memo), then memo[v] this will not run
        memo.append(memo[-1] + memo[-2])
    return memo[v]


# pows calculates powers of 2 for provided v and stores in memo
def pows(v, memo):
    if v not in memo:
        memo[v] = 1 << v  # Faster than pow(2, v)
    return memo[v]

def solution(A, B):
    combo_memo, mod_memo = [0,1,2,3,5], {}
    result = [0] * len(A)
    for i in range(len(A)):
        # For modulos with values of powers of 2, use bitwise AND (value-1)
        # Here, all in mod_memo are 2**n
        result[i] = combos(A[i], combo_memo) & (pows(B[i], mod_memo)-1)
    return result


def test():
    cases = [
        {'input': ([4,4,5,5,1], [3,2,4,3,1]), 'expected': [5,1,8,0,1]},
        {'input': ([1,1,1,1,1], [1,1,1,1,1]), 'expected': [1,1,1,1,1]},
        {'input': ([1,1,1,1,1], [2,2,2,2,2]), 'expected': [1,1,1,1,1]},
        {'input': ([1,1,1,1,1], [3,3,3,3,3]), 'expected': [1,1,1,1,1]},
    ]
    for c in cases:
        if solution(*c['input']) != c['expected']:
            print("Failed test with input {}. Got {}, want {}".format(c['input'], solution(*c['input']), c['expected']))
            return
    print('All tests passed')

test()

File: codility/go/13-ladder.go
package golang

func ladder(A, B []int) []int {
	// number of combinations to traverse staircase match numOfSteps = fibAtPosition+1
	// we need to find a fib number that is one higher than the number of steps
	var maxSteps int
	for _, a := range A {
		if a > maxSteps {
			maxSteps = a
		}
	}

	// there are only two possible ways to make a step, so this will match the fibonacci sequence
	// can be easily extended to 3+: use i-3 + i-2 + i-1
	// we later want to use A[numOfSteps] to get the result, plus we'll use one helper item for fib sequence
	combos := make([]int, maxSteps+2)
	combos[0], combos[1] = 0, 1
	for i := 2; i < len(combos); i++ {
		combos[i] = combos[i-2] + combos[i-1]
	}

	// the trick to speeding up the calculation is to use bitwise AND instead od modulo for large numbers
	// remove helper item from fibs
	combos = combos[1:]
	// memo powers of two as length of B can be up to 50000 with only 30 distinct numbers, per task requirement
	modMemo := make(map[int]int64)
	result := make([]int, len(A))
	for i := range result {
		if _, ok := modMemo[B[i]]; !ok {
			modMemo[B[i]] = 1 << B[i]
		}

		combos := combos[A[i]]
		m := modMemo[B[i]] - 1
		result[i] = int(int64(combos) & m)
	}

	return result
}
